### ISSUE TITLE
Producer.committableSink doesn't respect commit settings or apply backpressure

### ISSUE THREAD

**Issue by gabrielreid** [akka/alpakka-kafka#1053]
### Versions used 

alpakka-kafka 2.0.1

### Expected Behavior

When using `Producer.committableSink(producerSettings, committerSettings)`, it is expected that the provided commit settings be respected, and that the progress of the stream be dependent on the progress of commits being executed (i.e. that backpressure be applied upstream based on the speed with which commits are completed).

For example, if using committer settings with max batch size of 1 and parallelism of 1 like so

    CommitterSettings(system).withMaxBatch(1L).withParallelism(1)

within `Producer.committableSink`, it is expected that the upstream stream will only progress one message at a time, as commits complete.

### Actual Behavior

When using this particular overload of `Producer.committableSink` (which internally uses `CommittingProducerSinkStage`), it appears that the batch size and parallelism settings within the provided `CommitterSettings` are basically ignored (or misinterpreted), and the upstream rate is unaffected by the rate with which commits are completed.

As a somewhat extreme example, consider the above `CommitterSettings` with `maxBatch` and `parallelism` both set to 1. This will result in every passed-through message resulting in an internal commit message being sent to the `KafkaConsumer`. However, the returned `Future` from this commit call is not specifically waited upon (it's just used to count down the number of in-flight commits). 

Internally, the `KafkaConsumer` does its own buffering of commits based on the number of commit messages that it receives between its own internal Kafka polling loop, and this internal buffering doesn't have any knowledge of the `CommitterSettings`, meaning that much larger buffers of commits can be performed, regardless of the `maxBatch` settings.

This can also result in many concurrent in-flight commit requests regardless of the `parallelism` setting, for the same reasons as explained above.

The above factors also mean that the stream is not backpressured at all based on the speed with which commits can be handled. Due to all of this, we've seen that it can be quite easy to run into commit timeouts when using this variant of `Producer.committableSink`, as it appears that Kafka gets somewhat overloaded with commit requests (also probably due to the fact that each in-flight commit request is waiting on an ack from Kafka).

This issue does not appear to affect the deprecated overloads of `Producer.committableSink` which make use of `Producer.flexiFlow`, as the commits are a strictly down-stream process and run within `mapAsync`.

### Reproducible Test Case

I'm able to create a reproducible test case quite easily where I can see the result of this issue in output logging, but at the moment I don't see any easy way to actually replicate the underlying problem in a unit test.

**Comment by ennru** [comment]
Thank you @gabrielreid for pointing this out. 

Not taking commit callbacks into account for the backpressure is by design and desired behaviour. But I agree, it could be documented. Not all `CommitterSettings` are used by the new `committableSink` implementation.

The backpressure will be applied when the Kafka producer starts blocking on full buffers.

We need to understand why the commit timeouts you notice arise. With the aggregation of commits in the Kafka consumer actor introduced in 1.1.0, we avoid overloading the Kafka consumer with commit requests, there must be something else to it.

**Comment by gabrielreid** [comment]
Thanks for the explanation @ennru 

What's maybe quite unusual about our situation is that the majority of the incoming consumer messages are simply translated into `ProducerMessage.passThrough`, meaning that there is usually nothing being sent to the Kafka producer, and so no backpressure occurring from there, which leads to the huge number of concurrent in-flight commits (and then commit timeouts).

It seems likely that it would just be better in this specific situation to use one of the deprecated overloads of `Producer.committableSink`. Probably documenting the trade-offs of this particular `committableSink` overload as well as maybe removing the deprecation from one of the currently-deprecated overloads would be sufficient.

**Comment by seglo** [comment]
@gabrielreid `Producer.committableSink` was recently modified under the hood to use the new `CommittingProducerSinkStage` which optimizes the produce and committing logic.  If you want back-pressure on offset commits you can still do it using `Committer.sink` with a `CommitDelivery` setting of `WaitForAck` (which is the default).

**Comment by gabrielreid** [comment]
@seglo thanks for pointing that out. 

It would indeed work for our specific use case to use `Producer.flexiFlow` combined with `Committer.sink`, using `WaitForAck`.

I think that the main issue here was that we were running into lots of commit timeouts because of our heavy use of `ProducerMessage.passThrough`, which meant that there was no throttling at all on the `Producer.committableSink`, which resulted in a huge number of concurrent in-flight commit requests.

Another issue here is also that the semantics of the `akka.kafka.committer.max-batch` change depending on which `Producer.committableSink` overload is chosen:
* Using the deprecated overloads, the `max-batch` does what it says, i.e. it batches the commits sent to Kafka as specified
* Using the newer `Producer.committableSink` overload (based on `CommittingProducerSinkStage`), the `max-batch` setting specifies the batching of commits to the internal `KafkaConsumerActor`, but the actor itself then does its own batching of commits to Kafka that is independent of the `max-batch` setting

I can make a PR to update the docs to reflect the information in this ticket, i.e.
* that backpressure is not applied on commits using the newer `Producer.committableSink` overload
* clarification on the different semantics of `akka.kafka.committer.max-batch` depending on where it's used
* the suggestion to use `Producer.flexiFlow` combined with `Committer.sink` if you want to join producing and committing in a flow which makes heavy use of pass-through messages that don't result in the Kafka Producer being used

My main concern with just documenting this is that it feels like it's quite out of line with the [core design principles of Akka Streams](https://doc.akka.io/docs/akka/current/general/stream/stream-design.html#what-shall-users-of-akka-streams-expect-), in that it implies quite a few exceptions to rules that need to be remembered.

That being said, having this behaviour documented is still a big improvement over it not being documented and requiring that it be discovered via trial and error.

Shall I make the PR for the doc update? Or is it preferable (or possible) to find a way to make things more stable/reliable for the use case where lots of pass-through messages are being used?

**Comment by ennru** [comment]
I wonder if we could come up with something that would support your use with the `CommittingProducerSinkStage`.
Can you give an indication of what numbers we're talking about? 
Do you commit with `max-batch = 1000`, how many messages?
Default poll interval (50 ms)?

We could have another counter in the actor which either counts received `Commit` messages, or gets the number of elements that lead to the commits from the commit batches. When that counter passes a limit it would `requestDelayedPoll()`.

**Comment by gabrielreid** [comment]
We were using `max-batch = 25`, with all the rest left as defaults when I first started investigating this issue. The rate of message processing would have been around 500 per second, with nearly all outgoing envelopes being passThrough messages (meaning that the Producer didn't have to do anything for them).

Due to a change in employment, as of today I'm no longer working on the codebase that was demonstrating this issue, although I'd be happy to help try to resolve it if possible.

@ennru that approach of keeping count of things in the actor sounds interesting. I was looking for some way of handling this in `CommittingProducerSinkStage` itself, but maybe doing it in the actor is indeed the way to go.

**Comment by ennru** [comment]
What was the idea behind `max-batch = 25`? 
With the new aggregation of commits in the consumer actor, this only leads to more actor messages being sent to the consumer actor, and more promises/futures to keep track of.

The question is what to count in the consumer actor. Right now the aggregation of the received offsets is deferred to just before the commit happens. So it can't easily see the number of relevant offsets that will be committed. We should possibly send more information about what triggered the `Commit` message.
This is a bit related to https://github.com/akka/alpakka-kafka/pull/1058 which looks into adding some flushing of the commits when errors occur.

**Comment by gabrielreid** [comment]
> What was the idea behind `max-batch = 25`?
> With the new aggregation of commits in the consumer actor, this only leads to more actor messages being sent to the consumer actor, and more promises/futures to keep track of.

Indeed, the use of such a low `max-batch` appeared to be the cause of the issue (or in other words, the issue could be avoided by increasing the `max-batch`). We were using such a low `max-batch` purely due to the fact that it _looked_ like it would do what we wanted, namely offer much better performance than a `max-batch` of 1, while also having something that came much closer to the guarantees offered by `max-batch = 1` (i.e. a much smaller number of "lost" commits in the event of a failure somewhere). This was done based on the assumption that there would be backpressure on the commits, and without the knowledge that it would just result in a huge number of in-flight commits.

As mentioned earlier, using this setting basically didn't work at all reliably, which is what lead me to report this issue :-)

> The question is what to count in the consumer actor. Right now the aggregation of the received offsets is deferred to just before the commit happens. So it can't easily see the number of relevant offsets that will be committed. We should possibly send more information about what triggered the `Commit` message.

Would it work to just increment and decrement a counter of the currently in-flight `commitAsync` calls in `KafkaConsumerActor#commit`, and trigger the `requestDelayedPoll()` based on that counter?

**Comment by ennru** [comment]
> Would it work to just increment and decrement a counter of the currently in-flight `commitAsync` calls in `KafkaConsumerActor#commit`, and trigger the `requestDelayedPoll()` based on that counter?

That is one way of doing it. The alternative would be to actually report how many consumed elements that went into the commit message (which is closer to what the `max-batch` used to mean before v1.1).
